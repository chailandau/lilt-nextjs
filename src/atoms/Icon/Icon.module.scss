.sm {
    width: rem(16);
    height: rem(16);
}
.md {
    width: rem(24);
    height: rem(24);
}
.lg {
    width: rem(32);
    height: rem(32);
}
.gray {
    fill: color(neutral, mdGray);
}
.green {
    fill: color(blueTheme, neonGreen);
}
.blue {
    fill: color(blueTheme, neonBlue);
}

.green-theme {
    &.green {
        fill: color(greenTheme, neonGreen);
    }
    &.blue {
        fill: color(greenTheme, neonBlue);
    }
}
