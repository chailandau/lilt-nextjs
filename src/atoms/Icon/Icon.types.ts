export const iconValues = ['caret'] as const;

export type iconIds = (typeof iconValues)[number];
