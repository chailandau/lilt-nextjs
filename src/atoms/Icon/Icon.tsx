import classNames from 'classnames';
import { m } from 'framer-motion';
import React, { type FC, Ref, forwardRef } from 'react';

import type { iconIds } from './Icon.types';

import styles from './Icon.module.scss';
import { iconValues } from './Icon.types';

export const iconColors = ['blue', 'green', 'gray'] as const;

export const iconSizes = ['sm', 'md', 'lg'] as const;

export interface IconProps {
    /** Icon id */
    id: iconIds;
    /** Icon size */
    size?: (typeof iconSizes)[number];
    /** Icon fill color */
    color?: (typeof iconColors)[number];
    /** Forwarded ref (optional) */
    ref?: Ref<SVGSVGElement>;
}

const Icon: FC<IconProps> = forwardRef(
    ({ id, size = 'sm', color = 'gray' }, ref) => {
        if (!iconValues.includes(id)) {
            return null;
        }
        const classList = classNames(styles[color], styles[size]);

        return (
            <svg className={classList} ref={ref}>
                <use href={`/icons/icon-sprites.svg#${id}`} />
            </svg>
        );
    }
);

export default Icon;

export const MotionIcon = m(Icon);
