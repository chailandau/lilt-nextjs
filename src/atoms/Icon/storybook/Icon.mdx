import { Canvas, Meta, ArgTypes } from '@storybook/blocks';

import * as IconStories from './Icon.stories';

<Meta of={IconStories} />

# Icon Atom

<Canvas of={IconStories.Default} />

<br />

### id

The icon `id` is used to select which icon to display.

<hr />

### size

Select a size for the `Icon` atom. If no size is selected, size will default to `sm`.

<hr />

### color

Select a color for the `Icon` atom's fill. If no color is selected, color will default to `gray`.

<hr />

### ref

An optional ref that is forwarded to the underlying SVG element.

<br />
<br />
<br />

## Props

<ArgTypes />
