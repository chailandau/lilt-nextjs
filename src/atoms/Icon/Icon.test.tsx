import { composeStory } from '@storybook/react';
import { render } from '@testing-library/react';

import { MotionIcon, iconColors, iconSizes } from './Icon';
import Meta, { Default, Large, Medium } from './storybook/Icon.stories';

import {
    testAxeViolations,
    testColors,
    testMatchesSnapshot,
    testSizes
} from '@/utils/testHelpers';

const getIconInfo = (Variant: typeof Default) => ({
    variant: composeStory(Variant, Meta),
    size: Variant.args?.size || ''
});

const iconVariants = [
    getIconInfo(Default),
    getIconInfo(Medium),
    getIconInfo(Large)
];

describe('Icon', () => {
    const Icon = composeStory(Default, Meta);
    testSizes(<Icon />, iconSizes);
});

iconVariants.forEach(({ variant: Icon, size }) => {
    describe(size, () => {
        it("doesn't render if `id` is not valid", () => {
            const { container } = render(
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore
                <Icon id='test' color={undefined} size={undefined} />
            );
            const svg = container.querySelector('svg');
            expect(svg).toBeNull();
        });
        test('renders MotionIcon component', () => {
            const { container } = render(<MotionIcon id='caret' />);
            const svg = container.querySelector('svg');
            expect(svg).toBeDefined();
        });
        testColors(<Icon />, iconColors);
        testAxeViolations(<Icon />);
        testMatchesSnapshot(<Icon />);
    });
});
