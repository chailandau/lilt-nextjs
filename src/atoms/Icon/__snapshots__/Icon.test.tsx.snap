// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon has 'lg' class 1`] = `
<svg
  className="gray lg"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`Icon has 'md' class 1`] = `
<svg
  className="gray md"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`Icon has 'sm' class 1`] = `
<svg
  className="gray sm"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`lg has 'blue' class 1`] = `
<svg
  className="blue lg"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`lg has 'gray' class 1`] = `
<svg
  className="gray lg"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`lg has 'green' class 1`] = `
<svg
  className="green lg"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`lg matches snapshot 1`] = `
<svg
  className="gray lg"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`md has 'blue' class 1`] = `
<svg
  className="blue md"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`md has 'gray' class 1`] = `
<svg
  className="gray md"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`md has 'green' class 1`] = `
<svg
  className="green md"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`md matches snapshot 1`] = `
<svg
  className="gray md"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`sm has 'blue' class 1`] = `
<svg
  className="blue sm"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`sm has 'gray' class 1`] = `
<svg
  className="gray sm"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`sm has 'green' class 1`] = `
<svg
  className="green sm"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;

exports[`sm matches snapshot 1`] = `
<svg
  className="gray sm"
>
  <use
    href="/icons/icon-sprites.svg#caret"
  />
</svg>
`;
